charmtools/build/builder.py

```diff
@@ -117,8 +117,14 @@ def fetch(self):
                 self.fetched_url = getattr(fetcher, "fetched_url", None)
                 self.vcs = getattr(fetcher, "vcs", None)
             self.revision = fetcher.get_revision(self.directory)
-            self.branch = fetcher.get_branch_for_revision(self.directory,
-                                                          self.revision)
+            # NOTE(ajkavanagh): Due to bug:#606 there isn't a good way to
+            # determine the branch for a revision as there isn't a one-to-one
+            # correspondence.  For the moment, work-around the bug by just
+            # using the revision as the branch so that things do at least
+            # build.
```

Comment:
I don't actually understand the use-case for "locking" to something mutable like a branch, but if you did want to do this, then you probably want to choose the non-`main`/`master` branch whose head exactly matches the revision, if any, since if you just cloned the repo and got want to be tracking a branch, you should be exactly on that branch at that point. (If my assumption there isn't correct, you could also compute the "closest" branch, i.e. least number of commits delta.) There's still the possibility of ambiguity, but I would expect it to be a lot more rare.